synapse/config/federation.py

```diff
@@ -21,6 +21,19 @@
 from synapse.config._util import validate_config
 
 
+DEFAULT_IP_RANGE_BLACKLIST = [
```

Comment:
I'm not 100% sure where these came from? They seem to partially block "private" IP spaces, but not all, see https://github.com/netaddr/netaddr/blob/master/netaddr/ip/__init__.py#L1918-L1972 / [RFC6890](https://tools.ietf.org/html/rfc6890).

Reply:
I think it was copied from the url_preview blacklist, which itself has evolved rather than being put together carefully (see https://github.com/matrix-org/synapse/pull/1198 for example). It might be a good time to consider if there are others we should add.

Reply:
In addition to what is there we should probably add:

* `192.0.0.0/24`: IANA IPv4 Special Purpose Address Registry (RFC 5736)
* `198.18.0.0/15`: Testing of inter-network communications between subnets (RFC 2544)
* `0.0.0.0/8`: Broadcast message (RFC 1700)
* `192.0.2.0/24`: TEST-NET examples and documentation (RFC 5737)
* `198.51.100.0/24`: TEST-NET-2 examples and documentation (RFC 5737)
* `203.0.113.0/24`: TEST-NET-3 examples and documentation (RFC 5737)

I don't see why we would ever talk to multicast addresses:

* `239.0.0.0` - `239.255.255.255`: Administrative Multicast
* `224.0.0.0/4`: Multicast
* `240.0.0.0/4`: Reserved for  multicast assignments (RFC 5771)
* `233.252.0.0/24`: Multicast test network
* `234.0.0.0` - `238.255.255.255`: Reserved multicast
* `225.0.0.0` - `231.255.255.255`: Reserved multicast
* `ff00::/8`: Multicast

Ones we should modify:
* `fe80::/64` -> `fe80::/10`: link local

Ones I'm unsure about:

* `192.88.99.0/24`: 6to4 anycast relays (RFC 3068)
* `fec0::/10`: Site Local Addresses (deprecated - RFC 3879)
* The other IPv6 reserved:  `ff00::/12`, `::/8`, `0100::/8`, `0200::/7`, `0400::/6`, `0800::/5`, `1000::/4`, `4000::/3`, `6000::/3`, `8000::/3`, `A000::/3`, `C000::/3`, `E000::/4`, `F000::/5`, `F800::/6`, `FE00::/9`

That's quite a big change though, what do you think?

Reply:
generally I agree with adding all of those to the default value, but a few points:

*  Note that `0.0.0.0/32` is already special-cased (I can't quite remember the history there). Should we update that special-casing rather than add `0.0.0.0/8` to the default here?
* Are the IPv6 reserved addresses just "not yet assigned"? My instinct is not to blacklist them (though I'm surprised to see `::/8` there, given `::1` is localhost)

Reply:
I believe you're correct that those IPv6 addresses have just not been assigned yet, which is quite different.

Reply:
Seems like https://tools.ietf.org/html/rfc6890 is the proper reference to be using.

Reply:
The list mentioned above looks sane to me.

IMO, we could and should add everything from the above lists, up to the "not yet assigned" addresses.

Regarding Site Local Addresses, RFC 3879 mentions that the block is not supposed to be reassigned except by future IETF action and that router implementations SHOULD prevent routing. Taking that into account, I don't see any downside in adding them to the default blacklist and it's nice from a defence in depth perspective, just in case there's a bad/old router implementation somewhere.

I think we can safely skip the "not yet assigned" addresses for now. On the topic of `::1`, I suppose the intent is that the entire `::/8` block is unassigned, except for `::1` which is treated as an exception and a special address.

I can't think of anything else that should be added to the list at the moment.